Explore: Fix unsubscribing from Loki websocket

aocenas opened the pull request:
Closes: https://github.com/grafana/grafana/issues/19255

Seems like `takeUntil` does not unsubscribe from the notifier when the main observable when it finishes which seems pretty weird to me but `take(1)` on the notifier fixes that.

One problem this creates though is that now 2 websocket connections are created. One main and second for the notifier in `takeUntil` this seems also strange to me as the observable is `multicast()`ed and `refCount()`ed in the datasource so still need to figure that out. @hugohaggmark if you have any tips here please let me know.

aocenas committed:
Fix unsubscribe

torkelo commented:
This is super strange, tested that unsubscribe on the datasource returned observable was called after I added that

aocenas commented:
> This is super strange, tested that unsubscribe on the datasource returned observable was called after I added that

Probably depends where on the chain the unsubscribe was called, because it was called but as there were 2 subscriptions it went only up to the `multicast` [inside the liveStream](https://github.com/grafana/grafana/blob/991f77cee1da71af12d3e853a152537420eb90d7/public/app/plugins/datasource/loki/live_streams.ts#L44) so there was still one subscription on the websocket itself.

In any case would definitely like to have another set of eyes on this as looking at the code I would also expect this to work before. Maybe the real error is somewhere else and this just a workaround.

hugohaggmark reviewed (commented):
I'm looking at this as well this morning. If I remove the `caching` in live stream and remove the `merge` from runRequests (see below) this should just work IMO as every request will get a new websocket stream and there should only be cold observables left in the pipeline. And it does work except for one scenario I found. Maybe we should sync after daily today @aocenas @torkelo?

1. Using the code below and using a Dashboard with two Panels with the same `live` loki query.
2. Edit one of them and turn off `live`.
3. There is just one live websocket (Correct) but the Panel that is still live doesn't update when i go back to dashboard. 
4. But if I save and refresh the browser, it works as expected. Seems like the transition from `Edit` to `View` doesn't play well for live streaming Panels but this is another bug I believe.

**removing cache**
```typescript
export class LiveStreams {
  getStream(target: LiveTarget): Observable<DataFrame[]> {
      const data = new CircularDataFrame({ capacity: target.size });
      data.labels = parseLabels(target.query);
      data.addField({ name: 'ts', type: FieldType.time, config: { title: 'Time' } });
      data.addField({ name: 'line', type: FieldType.string });
      data.addField({ name: 'labels', type: FieldType.other });

      const stream = webSocket(target.url).pipe(
        map((response: LokiResponse) => {
          appendResponseToBufferedData(response, data);
          return [data];
        }),
      );
    return stream;
  }
}
```

**removing merge**
```typescript
export function runRequest(datasource: DataSourceApi, request: DataQueryRequest): Observable<PanelData> {
  let state: RunningQueryState = {
    panelData: {
      state: LoadingState.Loading,
      series: [],
      request: request,
    },
    packets: {},
  };

  // Return early if there are no queries to run
  if (!request.targets.length) {
    request.endTime = Date.now();
    state.panelData.state = LoadingState.Done;
    return of(state.panelData);
  }

  const dataObservable = callQueryMethod(datasource, request).pipe(
    // Transform response packets into PanelData with merged results
    map((packet: DataQueryResponse) => {
      if (!isArray(packet.data)) {
        throw new Error(`Expected response data to be array, got ${typeof packet.data}.`);
      }

      request.endTime = Date.now();

      state = processResponsePacket(packet, state);
      return state.panelData;
    }),
    // handle errors
    catchError(err =>
      of({
        ...state.panelData,
        state: LoadingState.Error,
        error: processQueryError(err),
      })
    ),
    // finalize is triggered when subscriber unsubscribes
    // This makes sure any still running network requests are cancelled
    finalize(cancelNetworkRequestsOnUnsubscribe(request)),
  );

  return dataObservable;
}
```

hugohaggmark commented:
I've tested some more and I think we can get rid of the multicast bit in `live_streams.ts`. I think the `WebSocketSubject` handles subscribers automagically or perhaps I'm missing a use case here. Any thoughts @ryantxu ?

```typescript
export class LiveStreams {
  private streams: KeyValue<Observable<DataFrame[]>> = {};

  getStream(target: LiveTarget): Observable<DataFrame[]> {
    let stream = this.streams[target.url];
    if (!stream) {
      const data = new CircularDataFrame({ capacity: target.size });
      data.labels = parseLabels(target.query);
      data.addField({ name: 'ts', type: FieldType.time, config: { title: 'Time' } });
      data.addField({ name: 'line', type: FieldType.string });
      data.addField({ name: 'labels', type: FieldType.other });

      // const subject = new BehaviorSubject<DataFrame[]>([]);
      stream = webSocket(target.url).pipe(
        finalize(() => {
          delete this.streams[target.url];
        }),
        map((response: LokiResponse) => {
          appendResponseToBufferedData(response, data);
          return [data];
        }),
      );
      this.streams[target.url] = stream;
    }
    return stream;
  }
}
```

aocenas commented:
@hugohaggmark thanks for looking at this. The `multicast` in LiveStreams was there for the BehaviourSubject so that when connecting to existing webSocket it will get the last buffer update even if it was received before the subscription. So if you create a new panel with the same websocket it will show some data even if no new data was received.

hugohaggmark commented:
> @hugohaggmark thanks for looking at this. The `multicast` in LiveStreams was there for the BehaviourSubject so that when connecting to existing webSocket it will get the last buffer update even if it was received before the subscription. So if you create a new panel with the same websocket it will show some data even if no new data was received.

Ok, I see thats why it's there. Thanks

aocenas committed:
Remove behavior subject

aocenas commented:
So used @hugohaggmark s fix here. Also removed my hacked together debugSubscription and added dev dependency on https://github.com/cartant/rxjs-spy which is pretty cool and gives much more features for debugging.

aocenas commented on public/app/plugins/datasource/loki/live_streams.test.ts:
```diff
@@ -37,10 +37,9 @@ describe('Live Stream Tests', () => {
+    expect.assertions(4);
 
     const tests = [
-      (val: DataFrame[]) => expect(val).toEqual([]),
```
This is due to removal of behaviour subject.

torkelo reviewed (approved):
Looks good!
